pallets/worker/src/lib.rs

```diff
@@ -397,29 +400,73 @@ decl_module! {
             }
 
             let mut stream = RlpStream::new();
             Self::rlp_append(header.clone(), &mut stream);
             let rlp_header: Vec<u8> = stream.out();
 
             let signer = Signer::<T, T::AuthorityId>::any_account();
-
+            let dataset_info = StorageValueRef::persistent(b"light-client-worker::ethash-dataset");
+            let dataset = match dataset_info.get::<Vec<u8>>() {
+                Some(dataset) => Some(dataset.unwrap()),
+                None => {
+                    // FIXME(shekohex): move this code to a function guarded by std feature.
+                    // so we can only generate the dataset on std supported targets.
+                    debug::native::info!("Generate dataset for stored epoch: {}", stored_epoch);
+                    // lock the storage.
+                    let mut lock = StorageLock::<Time>::new(b"light-client-worker::ethash-dataset");
+                    let maybe_dataset = if let Ok(_guard) = lock.try_lock() {
+                        let dataset_size = ethash::get_full_size(stored_epoch as usize);
+                        let mut dataset = Vec::with_capacity(dataset_size);
+                        dataset.resize(dataset_size, 0u8);
+                        ethash::make_dataset(&mut dataset, &cache);
+                        dataset_info.set(&dataset);
+                        Some(dataset)
+                    } else {
+                        // no work to be done here, until the dataset generation is complete.
+                        None
+                    };
+                    maybe_dataset
+                },
+            };
+            let dataset = match dataset {
+                Some(dataset) => dataset,
+                // this skips the current invocation, as there is another worker is generating the
+                // dataset.
+                None => return
+            };
+            let should_generate_dataset = should_generate_dataset(U256::from(header.number), stored_epoch);
+            debug::native::info!("Should we generate the dataset? {}", should_generate_dataset);
+            if should_generate_dataset {
+                // FIXME(shekohex): lock the storage here to avoid regeneation.
+                // not yet sure how we will swap these and when.
+                let dataset_size = ethash::get_full_size(epoch as usize);
+                let mut dataset = Vec::with_capacity(dataset_size);
+                dataset.resize(dataset_size, 0u8);
+                ethash::make_dataset(&mut dataset, &cache);
+                // TODO(shekohex): save the dataset.
+            }
             let latest_block_number = Self::last_block_number();
             let call = if U256::from(header.number) > latest_block_number {
                 if Self::initialized() {
-                    // TODO: Add proof generation/fetching
+                    // TODO(drewstone): we should convert this proofs to DoubleNodeWithMerkleProof
```

Comment:
Yes we'll want to convert this to DoubleNodeWithMerkleProof. If we get the same output as the ethashproofs then it should be easy,
- we just use https://github.com/edgeware-builders/substrate-eth-light-client/blob/master/pallets/worker/src/tests.rs#L174 to encode and decode
- then we use https://github.com/edgeware-builders/substrate-eth-light-client/blob/master/pallets/worker/src/tests.rs#L246

Reply:
Question is whether this will work in WASM but I'm sure there's a hex decoder library for no_std.